### PR TITLE
added demo code to review

### PR DESCRIPTION
**Pull request by ashtilawat23**
It is very unlikely that we will accept a pull request for this repo. See the [README](https://github.com/dotnet/ef6/blob/main/README.md) for more info.

### CODE REVIEW

File: review/Product.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+using System;
```
Comment (ashtilawat23) [inline review/Product.cs:1]: Here is the review of the given code in `review/Product.cs`:

---

### Issue 1:

**Location:** `review/Product.cs:29`  
**Problem:** The `ApplyDiscount` method does not handle percentages greater than 100 correctly. While the method correctly throws an exception for percentages out of bounds, it fails to prevent reducing the `Price` below zero for large discounts close to 100.

**Fix:**
```csharp
// Before
if (percentage < 0 || percentage > 100)
    throw new ArgumentException("Discount percentage must be between 0 and 100");

Price = Price - (Price * percentage / 100);

// After
if (percentage < 0 || percentage > 100)
    throw new ArgumentException("Discount percentage must be between 0 and 100");

decimal discountAmount = Price * percentage / 100;
Price = Math.Max(0, Price - discountAmount);
```

**Impact:**
- Prevents the `Price` from becoming negative due to large discounts.
- Ensures product data integrity by maintaining valid price values.

**Testing:**
1. Attempt to apply a 100% discount and verify that `Price` is zero, not negative.
2. Test with valid discount percentages (e.g., 10%, 50%) to ensure correct price deduction.
3. Edge case: Apply discount when price is already zero and verify no errors occur.

---

### Issue 2:

**Location:** `review/Product.cs:42`  
**Problem:** No explicit check ensures the `Name` field is not null or empty, despite the `[Required]` attribute. While the attribute checks for null, it’s a good practice to include checks within the constructor or set method to ensure field integrity in all use cases.

**Fix:**
```csharp
// Before
public string Name { get; set; }

// After
private string _name;

[Required]
[StringLength(100)]
public string Name
{
    get => _name;
    set
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException("Product name cannot be null, empty, or whitespace.");
        _name = value;
    }
}
```

**Impact:**
- Ensures `Name` field is always set to a meaningful value.
- Avoids potential null reference exceptions or useless data states.

**Testing:**
1. Attempt to create a product with an empty or whitespace `Name` to ensure the validation error is raised.
2. Verify setting the `Name` with valid non-empty strings works as expected.
3. Test with null value assignment and ensure an exception is thrown.

---

### Issue 3:

**Location:** `review/Product.cs:19`  
**Problem:** `CreatedDate` is initialized with `DateTime.UtcNow`, which is a good practice, but its mutable nature could result in unintended modifications elsewhere in the code.

**Fix:**
```csharp
// Before
public DateTime CreatedDate { get; set; }

// After
public DateTime CreatedDate { get; private set; }
```

**Impact:**
- Protects against accidental or unauthorized modifications of the creation date after the object has been initialized.
- Ensures accurate and reliable tracking of product creation time.

**Testing:**
1. Verify that `CreatedDate` is set upon instantiation and cannot be modified later.
2. Attempt modifications through reflection or other means and ensure they are prevented.

---

Overall, the code follows good practices for modeling a basic product entity with data annotations for validation. Addressing the issues above would improve robustness, particularly in terms of data integrity and defensiveness against invalid operations.
